File: backend/app/services/rag/knowledge.py
"""Curated authoritative civic knowledge for grounded answers without external APIs."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class KnowledgeDoc:
    id: str
    title: str
    authority: str
    authority_level: str  # OFFICIAL | STATUTORY | COURT | TRUSTED_SECONDARY | UNKNOWN
    source_url: str
    content: str
    section: Optional[str] = None
    page: Optional[str] = None
    last_verified: str = "2024-01-01"
    state: Optional[str] = None
    document_type: str = "guidance"
    government_level: str = "national"
    language: str = "en"
    tags: list[str] = field(default_factory=list)


KNOWLEDGE_BASE: list[KnowledgeDoc] = [
    KnowledgeDoc(
        id="rti-act-2005-overview",
        title="Right to Information Act, 2005 — Overview",
        authority="Government of India / India Code",
        authority_level="STATUTORY",
        source_url="https://www.indiacode.nic.in/handle/123456789/2063",
        section="Preamble & Section 6",
        last_verified="2024-06-01",
        document_type="statute",
        government_level="national",
        tags=["rti", "information", "public authority", "application"],
        content=(
            "The Right to Information Act, 2005 provides a practical regime of right to information "
            "for citizens to secure access to information under the control of public authorities. "
            "Under Section 6, a person seeking information shall make a request in writing or through "
            "electronic means in English or Hindi or in the official language of the area, specifying "
            "particulars of the information sought. The Act does not require the applicant to give "
            "reasons for requesting the information. Fees and submission channels vary by public "
            "authority; applicants should verify the current fee and address from the concerned "
            "public authority or State/Central Information Commission guidance."
        ),
    ),
    KnowledgeDoc(
        id="rti-section-7-timeline",
        title="RTI Act Section 7 — Disposal of request",
        authority="Government of India / India Code",
        authority_level="STATUTORY",
        source_url="https://www.indiacode.nic.in/handle/123456789/2063",
        section="Section 7",
        last_verified="2024-06-01",
        document_type="statute",
        government_level="national",
        tags=["rti", "timeline", "30 days", "appeal"],
        content=(
            "Section 7 of the RTI Act provides that the Central Public Information Officer or State "
            "Public Information Officer shall, as expeditiously as possible, and in any case within "
            "thirty days of the receipt of the request, either provide the information on payment of "
            "such fee as may be prescribed or reject the request for any of the reasons specified in "
            "sections 8 and 9. Where information concerns life or liberty of a person, it shall be "
            "provided within forty-eight hours. If no decision is given within the time, it may be "
            "deemed a refusal, and the applicant may pursue the appeal process under the Act."
        ),
    ),
    KnowledgeDoc(
        id="cpgrams-overview",
        title="Centralized Public Grievance Redress and Monitoring System (CPGRAMS)",
        authority="Department of Administrative Reforms & Public Grievances (DARPG)",
        authority_level="OFFICIAL",
        source_url="https://pgportal.gov.in/",
        section="About CPGRAMS",
        last_verified="2024-08-01",
        document_type="portal",
        government_level="national",
        tags=["grievance", "complaint", "cpgrams", "escalation", "municipality"],
        content=(
            "CPGRAMS is the Government of India's online grievance redress mechanism. Citizens can "
            "lodge grievances related to service delivery by Central Government ministries/departments "
            "and participating organisations through https://pgportal.gov.in/. Users should keep the "
            "registration/reference number for tracking. For purely local municipal issues, citizens "
            "should first use the relevant State or Municipal Corporation grievance portal where "
            "available, and may escalate through higher authorities or State grievance systems as "
            "applicable. Exact escalation paths depend on the State and local body."
        ),
    ),
    KnowledgeDoc(
        id="tamil-nadu-municipal-grievance",
        title="Tamil Nadu municipal / urban local body grievance guidance",
        authority="Government of Tamil Nadu / Greater Chennai Corporation public services",
        authority_level="OFFICIAL",
        source_url="https://chennaicorporation.gov.in/",
        section="Citizen services / complaints",
        last_verified="2024-08-01",
        state="Tamil Nadu",
        document_type="portal",
        government_level="local",
        tags=["chennai", "tamil nadu", "drainage", "garbage", "roads", "municipality", "complaint"],
        content=(
            "For civic issues in Chennai such as drainage blockage, garbage, roads, streetlights, and "
            "water supply, citizens may lodge complaints with Greater Chennai Corporation through its "
            "official citizen services channels (including the corporation website and designated "
            "helplines/apps as published by GCC). Keep the complaint reference number and date. If a "
            "complaint remains unresolved, citizens may follow up with the ward/zone office, use the "
            "corporation grievance tracking facility, and where appropriate escalate through Tamil Nadu "
            "State grievance mechanisms or seek recorded information under the RTI Act from the "
            "concerned public authority. Always verify current submission channels on the official "
            "GCC/State portal before filing."
        ),
    ),
    KnowledgeDoc(
        id="municipal-services-general",
        title="Municipal civic service complaints — general process",
        authority="Ministry of Housing and Urban Affairs / Urban Local Bodies (general guidance)",
        authority_level="TRUSTED_SECONDARY",
        source_url="https://mohua.gov.in/",
        section="Urban governance — citizen services",
        last_verified="2024-05-01",
        document_type="guidance",
        government_level="national",
        tags=["municipality", "drainage", "garbage", "roads", "streetlights", "water", "complaint"],
        content=(
            "Urban local bodies (municipal corporations/municipalities) are generally responsible for "
            "local civic services such as drainage, solid waste, roads, street lighting, and local "
            "water distribution within their jurisdiction. Typical citizen steps: (1) identify the "
            "correct local body for the location; (2) lodge a complaint through the official municipal "
            "portal/helpline and note the reference number; (3) follow up if unresolved within the "
            "published service timelines; (4) escalate within the municipal hierarchy or State urban "
            "development/grievance channels as applicable; (5) where needed, request recorded "
            "information under RTI about complaint status, works sanctioned, or expenditure. Exact "
            "portals, fees, and timelines vary by State and city and must be verified from official "
            "sources for that jurisdiction."
        ),
    ),
    KnowledgeDoc(
        id="rti-drafting-practice",
        title="Good practice for drafting RTI requests",
        authority="Central Information Commission / RTI Act practice notes (summary)",
        authority_level="TRUSTED_SECONDARY",
        source_url="https://cic.gov.in/",
        section="Filing RTI applications",
        last_verified="2024-06-01",
        document_type="guidance",
        government_level="national",
        tags=["rti", "draft", "records", "public authority"],
        content=(
            "Effective RTI requests seek recorded information held by a public authority rather than "
            "opinions or explanations. Applicants should identify the public authority likely to hold "
            "the records, state the information sought clearly (e.g., copies of work orders, sanction "
            "amounts, dates of complaints and action taken), specify the relevant period, and indicate "
            "preferred format (inspection/copies). Do not invent PIO addresses or fee amounts; obtain "
            "them from the public authority's RTI disclosures or official portal. Mark drafts clearly "
            "as drafts until verified."
        ),
    ),
    KnowledgeDoc(
        id="consumer-protection-act-2019",
        title="Consumer Protection Act, 2019 — remedies for defective goods and services",
        authority="Government of India / India Code",
        authority_level="STATUTORY",
        source_url="https://www.indiacode.nic.in/handle/123456789/13624",
        section="Consumer rights & commissions",
        last_verified="2024-06-01",
        document_type="statute",
        government_level="national",
        tags=["consumer", "warranty", "refund", "defective", "consumer forum", "rights"],
        content=(
            "The Consumer Protection Act, 2019 protects consumers against defective products and "
            "deficient services. Consumers may first complain to the seller or manufacturer in writing. "
            "If the dispute is not resolved, consumers may approach the appropriate Consumer Disputes "
            "Redressal Commission based on the value of the claim, as prescribed under the Act and rules. "
            "Consumers should preserve invoices, warranty cards, and communication records."
        ),
    ),
    KnowledgeDoc(
        id="tenant-rights-general",
        title="Tenant and landlord disputes — security deposit and State tenancy rules",
        authority="Ministry of Housing and Urban Affairs / State tenancy frameworks (summary)",
        authority_level="TRUSTED_SECONDARY",
        source_url="https://mohua.gov.in/",
        section="Urban housing — tenancy",
        last_verified="2024-05-01",
        document_type="guidance",
        government_level="national",
        tags=["tenant", "landlord", "rent", "security deposit", "eviction", "lease", "rights"],
        content=(
            "Landlord-tenant disputes including non-return of security deposits are governed by applicable "
            "State rent control or tenancy laws and rental agreements. Tenants should keep the rent agreement, "
            "deposit receipts, and evidence of handover condition. Many States provide dispute resolution "
            "through rent control authorities or applicable forums; exact procedures vary by State and must "
            "be verified on the official State portal."
        ),
    ),
    KnowledgeDoc(
        id="labour-wage-complaints",
        title="Unpaid wages — labour department grievance channels",
        authority="Ministry of Labour & Employment / State labour departments (summary)",
        authority_level="TRUSTED_SECONDARY",
        source_url="https://labour.gov.in/",
        section="Wage protection",
        last_verified="2024-05-01",
        document_type="guidance",
        government_level="national",
        tags=["workplace", "employer", "salary", "wages", "labour", "rights"],
        content=(
            "Workers facing unpaid wages may raise the matter with the employer in writing and preserve "
            "employment records, payslips, and bank statements. State labour departments and labour "
            "commissioners provide official channels for wage-related complaints under applicable labour "
            "laws. Industrial dispute procedures may apply depending on the nature of employment; verify "
            "the correct State labour portal before filing."
        ),
    ),
    KnowledgeDoc(
        id="pm-kisan-eligibility",
        title="PM-KISAN — eligibility overview",
        authority="Ministry of Agriculture & Farmers Welfare",
        authority_level="OFFICIAL",
        source_url="https://pmkisan.gov.in/",
        section="Scheme guidelines",
        last_verified="2024-08-01",
        document_type="scheme",
        government_level="national",
        state="Tamil Nadu",
        tags=["pm-kisan", "pm kisan", "scheme", "eligibility", "farmer", "agriculture"],
        content=(
            "PM-KISAN provides income support to eligible landholding farmer families subject to scheme "
            "exclusion criteria published on https://pmkisan.gov.in/. Eligibility depends on land records, "
            "farmer family definition, and exclusion categories such as certain institutional landholders "
            "and income tax payers in higher brackets as notified. Applicants must register through the "
            "official PM-KISAN portal or authorised channels and verify State-specific land record requirements."
        ),
    ),
    KnowledgeDoc(
        id="rti-section-2h-public-authority",
        title="RTI Act Section 2(h) — definition of public authority (plain summary)",
        authority="Government of India / India Code",
        authority_level="STATUTORY",
        source_url="https://www.indiacode.nic.in/handle/123456789/2063",
        section="Section 2(h)",
        last_verified="2024-06-01",
        document_type="statute",
        government_level="national",
        tags=["bureaucracy", "public authority", "section 2(h)", "rti", "explain", "plain language"],
        content=(
            "Section 2(h) of the Right to Information Act, 2005 defines 'public authority' to mean any "
            "authority or body or institution of self-government established or constituted by or under "
            "the Constitution, by any law made by Parliament or a State Legislature, or by notification "
            "issued by the appropriate Government. It also includes bodies owned, controlled, or substantially "
            "financed by the Government. In simple terms: if a government department, municipality, or "
            "government-controlled body holds the records you need, it is likely a public authority for RTI."
        ),
    ),
    KnowledgeDoc(
        id="official-form-filling-guidance",
        title="Filling official government application forms — general guidance",
        authority="Digital India / government citizen services (summary)",
        authority_level="TRUSTED_SECONDARY",
        source_url="https://www.india.gov.in/",
        section="Citizen services",
        last_verified="2024-05-01",
        document_type="guidance",
        government_level="national",
        tags=["form", "application form", "fill", "prefill", "rti application", "complaint form"],
        content=(
            "Official government forms should be completed with accurate personal details, a clear subject, "
            "and specific information requested. For RTI forms, describe records sought rather than opinions. "
            "For grievance forms, include location, dates, and prior reference numbers if any. Always use "
            "the latest form version from the official portal and verify submission address and fee before filing."
        ),
    ),
]


def search_knowledge(
    query: str,
    *,
    state: Optional[str] = None,
    top_k: int = 5,
) -> list[KnowledgeDoc]:
    q = query.lower()
    tokens = [t for t in q.replace(",", " ").split() if len(t) > 2]

    scored: list[tuple[float, KnowledgeDoc]] = []
    for doc in KNOWLEDGE_BASE:
        score = 0.0
        hay = f"{doc.title} {doc.content} {' '.join(doc.tags)}".lower()
        for t in tokens:
            if t in hay:
                score += 1.0
        for tag in doc.tags:
            if tag in q:
                score += 2.0
        if state and doc.state and state.lower() in doc.state.lower():
            score += 3.0
        if "chennai" in q and "chennai" in hay:
            score += 3.0
        if "rti" in q and "rti" in doc.tags:
            score += 2.0
        if any(w in q for w in ("consumer", "tenant", "landlord", "employer", "salary", "warranty")):
            if any(t in doc.tags for t in ("consumer", "tenant", "workplace", "rights")):
                score += 2.5
        if any(w in q for w in ("eligible", "eligibility", "pm-kisan", "pm kisan", "scheme")):
            if any(t in doc.tags for t in ("scheme", "eligibility", "pm-kisan", "pm kisan")):
                score += 2.5
        if any(w in q for w in ("public authority", "section 2", "bureaucracy", "plain language", "explain")):
            if any(t in doc.tags for t in ("bureaucracy", "public authority", "plain language")):
                score += 2.5
        if any(w in q for w in ("form", "fill", "application form")):
            if any(t in doc.tags for t in ("form", "application form", "fill")):
                score += 2.0
        if any(w in q for w in ("drainage", "garbage", "municip", "complaint", "grievance")):
            if any(t in doc.tags for t in ("drainage", "municipality", "grievance", "complaint", "chennai")):
                score += 2.0
        if score > 0:
            # Prefer higher authority
            boost = {
                "STATUTORY": 1.5,
                "OFFICIAL": 1.3,
                "COURT": 1.2,
                "TRUSTED_SECONDARY": 1.0,
                "UNKNOWN": 0.2,
            }.get(doc.authority_level, 0.5)
            scored.append((score * boost, doc))

    scored.sort(key=lambda x: x[0], reverse=True)
    return [d for _, d in scored[:top_k]]

File: frontend/src/lib/civic/knowledge.ts
/** Curated civic knowledge for grounded answers without external APIs. */

import { CITY_DIRECTORY, statePortalUrl } from "./india-geography";

export interface KnowledgeDoc {
  id: string;
  title: string;
  authority: string;
  authority_level: string;
  source_url: string;
  content: string;
  section?: string;
  last_verified: string;
  state?: string;
  city?: string;
  domains: string[];
  tags: string[];
}

const CORE_DOCS: KnowledgeDoc[] = [
  {
    id: "rti-act-2005-overview",
    title: "Right to Information Act, 2005 — Overview",
    authority: "Government of India",
    authority_level: "STATUTORY",
    source_url: "https://rti.gov.in/",
    section: "Section 6",
    last_verified: "2024-06-01",
    domains: ["rti", "form_filler"],
    tags: ["rti", "information", "public authority", "application", "records"],
    content:
      "The Right to Information Act, 2005 allows citizens to request recorded information from public authorities. An RTI application must specify the particulars of information sought and the relevant public authority.",
  },
  {
    id: "rti-section-7-timeline",
    title: "RTI Act Section 7 — Disposal of request",
    authority: "Government of India",
    authority_level: "STATUTORY",
    source_url: "https://rti.gov.in/",
    section: "Section 7",
    last_verified: "2024-06-01",
    domains: ["rti"],
    tags: ["rti", "timeline", "30 days", "appeal", "disposal"],
    content:
      "Section 7 of the RTI Act provides that information shall ordinarily be provided within thirty days of receipt of the request. Life or liberty matters must be answered within forty-eight hours.",
  },
  {
    id: "rti-public-authority-2h",
    title: "RTI Act Section 2(h) — Public authority",
    authority: "Government of India",
    authority_level: "STATUTORY",
    source_url: "https://rti.gov.in/",
    section: "Section 2(h)",
    last_verified: "2024-06-01",
    domains: ["rti", "bureaucracy"],
    tags: ["rti", "public authority", "section 2(h)", "bureaucracy", "definition"],
    content:
      "A public authority under the RTI Act means any authority or body established under the Constitution, law, or government notification. Section 2(h) defines public authority and helps identify where to file an RTI application.",
  },
  {
    id: "consumer-protection-act",
    title: "Consumer Protection Act, 2019 — Remedies",
    authority: "Department of Consumer Affairs",
    authority_level: "STATUTORY",
    source_url: "https://consumerhelpline.gov.in/",
    last_verified: "2024-06-01",
    domains: ["rights_navigator"],
    tags: ["consumer", "refund", "warranty", "defective", "shop", "product"],
    content:
      "The Consumer Protection Act, 2019 provides remedies for defective goods and deficient services. Consumers should first raise a written complaint with the seller, preserve invoices and warranty documents, and approach the appropriate consumer commission if unresolved.",
  },
  {
    id: "national-consumer-helpline",
    title: "National Consumer Helpline",
    authority: "Department of Consumer Affairs",
    authority_level: "OFFICIAL",
    source_url: "https://consumerhelpline.gov.in/",
    last_verified: "2024-08-01",
    domains: ["rights_navigator"],
    tags: ["consumer", "helpline", "complaint", "refund", "warranty"],
    content:
      "The National Consumer Helpline allows consumers to register grievances about defective products and deficient services and seek guidance on consumer commission procedures.",
  },
  {
    id: "cpgrams-overview",
    title: "CPGRAMS — Central Public Grievance Portal",
    authority: "DARPG",
    authority_level: "OFFICIAL",
    source_url: "https://pgportal.gov.in/",
    last_verified: "2024-08-01",
    domains: ["grievance"],
    tags: ["grievance", "complaint", "central government", "pg portal", "escalation"],
    content:
      "CPGRAMS is the Centralized Public Grievance Redress and Monitoring System where citizens can lodge grievances related to Central Government ministries and departments and track status online.",
  },
  {
    id: "municipal-grievance-general",
    title: "Municipal grievance redress — general guidance",
    authority: "Urban local bodies",
    authority_level: "OFFICIAL",
    source_url: "https://pgportal.gov.in/",
    last_verified: "2024-06-01",
    domains: ["grievance"],
    tags: ["municipal", "drainage", "garbage", "road", "complaint", "local body"],
    content:
      "Urban local bodies are responsible for local civic services such as drainage, garbage, roads, and water within their jurisdiction. Citizens should lodge complaints through the official municipal portal or helpline and keep the complaint reference number.",
  },
  {
    id: "pm-kisan-eligibility",
    title: "PM-KISAN — eligibility overview",
    authority: "Ministry of Agriculture & Farmers Welfare",
    authority_level: "OFFICIAL",
    source_url: "https://pmkisan.gov.in/",
    last_verified: "2024-06-01",
    domains: ["scheme_eligibility"],
    tags: ["pm-kisan", "scheme", "eligibility", "farmer", "farmland", "agriculture"],
    content:
      "PM-KISAN provides income support to eligible landholding farmer families. Eligibility criteria including landholding limits and exclusion categories are published on the official PM-KISAN portal. Apply only through official government channels.",
  },
  {
    id: "labour-wages-india",
    title: "Unpaid wages — labour department redress",
    authority: "Ministry of Labour & Employment",
    authority_level: "OFFICIAL",
    source_url: "https://labour.gov.in/",
    last_verified: "2024-06-01",
    domains: ["rights_navigator"],
    tags: ["employer", "salary", "wages", "workplace", "labour", "unpaid", "termination"],
    content:
      "Unpaid wages and employment disputes may be addressed through State labour departments and labour commissioner channels. Workers should keep employment records, pay slips, and written communications as evidence.",
  },
  {
    id: "tenancy-general-india",
    title: "Landlord-tenant and deposit disputes — general guidance",
    authority: "State rent / tenancy laws (verify locally)",
    authority_level: "OFFICIAL",
    source_url: "https://labour.gov.in/",
    last_verified: "2024-06-01",
    domains: ["rights_navigator"],
    tags: ["tenant", "landlord", "deposit", "rent", "lease", "eviction", "security"],
    content:
      "Landlord-tenant disputes including security deposits are governed by applicable State rent and tenancy laws. Tenants should document the tenancy agreement, notice period, and deposit payment before approaching the relevant rent authority or consumer forum.",
  },
];

function cityDocs(): KnowledgeDoc[] {
  return CITY_DIRECTORY.map((entry) => ({
    id: `municipal-${entry.city.toLowerCase().replace(/\s+/g, "-")}`,
    title: `${entry.localAuthority} — citizen services`,
    authority: entry.localAuthority,
    authority_level: "OFFICIAL",
    source_url: entry.portalUrl || statePortalUrl(entry.state) || "https://pgportal.gov.in/",
    last_verified: "2024-06-01",
    state: entry.state,
    city: entry.city,
    domains: ["grievance", "rti", "form_filler"],
    tags: [
      entry.city.toLowerCase(),
      entry.state.toLowerCase(),
      "municipal",
      "complaint",
      "grievance",
      "local body",
      ...(entry.aliases || []),
    ],
    content: `For civic issues in ${entry.city}, ${entry.state}, citizens may lodge and track complaints through ${entry.localAuthority} official channels or the applicable State municipal portal.`,
  }));
}

export const KNOWLEDGE_BASE: KnowledgeDoc[] = [...CORE_DOCS, ...cityDocs()];

const DOMAIN_TAGS: Record<string, string[]> = {
  rti: ["rti", "information", "public authority", "records", "application"],
  grievance: ["grievance", "municipal", "complaint", "drainage", "garbage", "road", "local body"],
  rights_navigator: ["consumer", "tenant", "landlord", "employer", "salary", "wages", "refund", "deposit"],
  scheme_eligibility: ["scheme", "pm-kisan", "eligibility", "farmer", "agriculture"],
  form_filler: ["rti", "application", "form", "records"],
  bureaucracy: ["public authority", "section 2(h)", "bureaucracy", "rti", "definition"],
};

const RIGHTS_AREA_TAGS: Record<string, string[]> = {
  tenant: ["tenant", "landlord", "deposit", "rent", "lease"],
  consumer: ["consumer", "refund", "warranty", "defective", "shop"],
  workplace: ["employer", "salary", "wages", "labour", "workplace"],
};

export function searchKnowledge(
  query: string,
  options?: {
    state?: string;
    city?: string;
    domain?: string;
    area?: string;
    topK?: number;
  }
): KnowledgeDoc[] {
  const topK = options?.topK ?? 5;
  const q = query.toLowerCase();
  const tokens = q.split(/\s+/).filter((t) => t.length > 2);
  const domain = options?.domain || "";
  const area = options?.area || "";
  const domainTags = DOMAIN_TAGS[domain] || [];
  const areaTags = RIGHTS_AREA_TAGS[area] || [];

  const scored = KNOWLEDGE_BASE.map((doc) => {
    const hay = `${doc.title} ${doc.content} ${doc.tags.join(" ")} ${doc.city || ""} ${doc.state || ""}`.toLowerCase();
    let score = 0;

    for (const t of tokens) {
      if (hay.includes(t)) score += 1;
    }

    if (domain && doc.domains.includes(domain)) score += 4;
    for (const tag of domainTags) {
      if (doc.tags.includes(tag) || hay.includes(tag)) score += 2;
    }
    for (const tag of areaTags) {
      if (doc.tags.includes(tag) || hay.includes(tag)) score += 3;
    }

    if (options?.state && doc.state?.toLowerCase() === options.state.toLowerCase()) score += 3;
    if (options?.city && doc.city?.toLowerCase() === options.city.toLowerCase()) score += 5;

    return { doc, score };
  })
    .filter((x) => x.score > 0)
    .sort((a, b) => b.score - a.score);

  if (scored.length === 0) {
    const fallback = KNOWLEDGE_BASE.filter((d) => !domain || d.domains.includes(domain));
    return fallback.slice(0, topK);
  }

  return scored.slice(0, topK).map((x) => x.doc);
}

export function isValidOfficialUrl(url: string): boolean {
  return /^https:\/\/[a-z0-9.-]+\.[a-z]{2,}/i.test(url);
}
